runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -147,15 +154,16 @@ public Token appendCheckpoint() {
      *  @param snapshotTimestamp snapshot at which the checkpoint is taken.
      */
     @VisibleForTesting
-    public Token appendCheckpoint(Token snapshotTimestamp) {
+    public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
         long start = System.currentTimeMillis();
 
-        // If snapshotTimestamp (stream tail) is a negative value
+        // If stream tail is a negative value
         // we need to enforce a hole to ensure the start log address
         // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
         // Having a negative value will prevent trimming from happening.
-        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+        if (!Address.isAddress(streamTail)) {
             snapshotTimestamp = forceHole();
+            streamTail = snapshotTimestamp.getSequence();
         }
 
         try {
```

Comment:
(Seems I cannot select the next line for comment, so just using this line.)
```java
rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
```
Is this correct? snapshotTimestamp.getSequence() is the global tail, I think you need the stream tail in order to prevent the issue we used to address in #2072 right?

Reply:
My bad, this seems not related with issue #2072. And actually we have to read global tail to prevent sequencer regress.

Reply:
Exactly, because snapshotTimestamp (global tail) will always be greater or equal to the stream tail, we only need to read the global tail to prevent sequencer regression, i.e., reassignment of sequence number in a different epoch.

Reply:
@annym There are other issues with sequencer regress in current approach and I just talked with @Maithem . Let’s have a discussions tomorrow.